### ISSUE TITLE
Analyzer cannot handle Yarn2+ virtual packages

### ISSUE THREAD

**Issue by thirteenthstep** [oss-review-toolkit/ort#8058]
I'm currently working on a project in which the analyzer result will miss more than half of our dependencies. The problem can be easily reproduced with an almost empty project and a single dependency:
```json
{
  "name": "ort-test",
  "packageManager": "yarn@4.0.2",
  "dependencies": {
    "@react-native-community/netinfo": "11.2.1"
  }
} 
```
Executing `ort --info analyze -f JSON -o ./analyze-ort-test -i ./ort-test` will result in 0 packages found.

This Problem occures when:

1.  Yarn Berry (2+) is used
2. The dependency in question itself has peer dependencies

The dependency used by this example can be swapped with any other package with peer dependencies, e.g. `react-native` (depending on react).

When investigating the issue, it is likely caused by Yarn using a virtual file layer for packages with peer dependencies. If we have a look at the result of `yarn info -A -R --manifest --json` (the command used by ort to list deps) our dependency is listed using a "virtual" locator:

```json
{
  "value": "@react-native-community/netinfo@npm:11.2.1",
  "children": {
    "Instances": 1,
    "Version": "11.2.1",
    "Manifest": {
      "License": "MIT",
      "Homepage": "https://github.com/react-native-netinfo/react-native-netinfo#readme"
    }
  }
}
{
  "value": "ort-test@workspace:.",
  "children": {
    "Version": "0.0.0",
    "Manifest": {
      "License": null,
      "Homepage": null
    },
    "Dependencies": [
      {
        "descriptor": "@react-native-community/netinfo@virtual:7ef2ec639f3470d2191275e2f3fd8391c01d875228d71bf82ba9ce9a90858464fb499034540949cf9bef8723b3f2638ad6037a3f99ed4516ef466bf8a24771a3#npm:11.2.1",
        "locator": "@react-native-community/netinfo@virtual:7ef2ec639f3470d2191275e2f3fd8391c01d875228d71bf82ba9ce9a90858464fb499034540949cf9bef8723b3f2638ad6037a3f99ed4516ef466bf8a24771a3#npm:11.2.1"
      }
    ]
  }
}
```

Unfortunately all dependencies with "virtual" locators will be skipped by ort during dependency aggregation at [Yarn2.kt](https://github.com/oss-review-toolkit/ort/blob/979847bbb5a4558a7f8cbe2a7c5256600da913cb/plugins/package-managers/node/src/main/kotlin/Yarn2.kt#L616)

```kotlin
when {
    // Prevent @virtual dependencies because they are internal to Yarn.
    // See https://yarnpkg.com/features/protocols
    "virtual" in locatorType -> null 
```

I do not fully understand the code comment regarding the "internal" part, but if we prevent the skipping, e.g. by replacing the "virtual" part, the result seems to be correct:

```
Index: plugins/package-managers/node/src/main/kotlin/Yarn2.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/package-managers/node/src/main/kotlin/Yarn2.kt b/plugins/package-managers/node/src/main/kotlin/Yarn2.kt
--- a/plugins/package-managers/node/src/main/kotlin/Yarn2.kt    (revision 979847bbb5a4558a7f8cbe2a7c5256600da913cb)
+++ b/plugins/package-managers/node/src/main/kotlin/Yarn2.kt    (date 1702601223304)
@@ -591,7 +591,7 @@
      */
     private fun processDependencies(dependencies: JsonNode): List<Identifier> =
         dependencies.mapNotNull { dependency ->
-            val locator = dependency["locator"].textValue()
+            val locator = dependency["locator"].textValue().replace(Regex("(.+)@virtual:(.+)#(.+)"),"$1@$3")
             val locatorMatcher = EXTRACT_FROM_LOCATOR_PATTERN.matchEntire(locator)
             if (locatorMatcher == null) {
                 issues += createAndLogIssue(
```
 
As I do not know the intention behind the skipping I do not feel capable of suggesting a PR in this case, but I would be glad about any information regarding the logic behind it and potential side effects if we continue using this workaround?

**Comment by sschuberth** [comment]
Also, I wonder whether simply dropping the `when` clause and just keeping the `else` code would solve your issue, @thirteenthstep?

**Comment by thirteenthstep** [comment]
> Also, I wonder whether simply dropping the `when` clause and just keeping the `else` code would solve your issue, @thirteenthstep?

That should be the correct approach. Adjustments regarding the locator pattern matching would then be necessary to work with "virtual" locator strings.

I just tried changing `"(.+)@(\\w+):(.+)"` to `"(.+)@(\\w+).*:(.+)"`. This should capture the correct values, but for some reason this solution still results in round about 10% less dependencies then the previous workaround.

**Comment by sschuberth** [comment]
> > Also, I wonder whether simply dropping the `when` clause and just keeping the `else` code would solve your issue, @thirteenthstep?
> 
> That should be the correct approach.

Would you be willing to contribute a fix based on that, including any other needed changes plus ideally a test case, since you're in the best situation currently to reproduce the issue?

**Comment by thirteenthstep** [comment]
@sschuberth I'm not sure if it's actually possible to handle all variations. A closer look at those missing dependencies shows that there are many possible patterns for those "locator" entries, like
`typescript@patch:typescript@npm%3A5.0.2#optional!builtin<compat/typescript>::version=5.0.2&hash=b5f058`
or `@some-scope/some-package@virtual:29710f211c97df3f40f9aeae5805aee7882fc1494a5d32550cc951f29c81c0f54881d323519532f0d351bd92c1d19c52ff1ecddba1ffd0dfb608cc5779d3b644#npm:4.17.4::__archiveUrl=https%3A%2F%2Fsome.repo.url-4.17.4.tgz`.
Modifying the existing regex-approach appears to be quite challenging and I have no idea if there are even more possible variations/patterns. 
I wonder if it would be possible to take these information from `yarn.lock`instead?

**Comment by nnobelis** [comment]
@sschuberth 
@thirteenthstep 

Please don't take the skipping of "virtual" I put as golden: I have low Yarn expertise and at the time, it seems skipping made the most sense. I am fine with changing the logic.

**Comment by thirteenthstep** [comment]
Update: I could not come up with a solution to make it work for every scenario. Eventually we had to switch to [https://github.com/CycloneDX/cdxgen](https://github.com/CycloneDX/cdxgen) to analyze yarn4 dependencies.

**Comment by sschuberth** [comment]
@fviernau do you think this is stall an issue with your latest refactorings in mind?

**Comment by fviernau** [comment]
> [@fviernau](https://github.com/fviernau) do you think this is stall an issue with your latest refactorings in mind?

The latest refactorings have touched only `Npm`, `Pnpm` and `Yarn` but not `Yarn2`. `Yarn2` has already been a separate implementation. However, I'm also planning to clean-up the `Yarn2` implementation and remove some code redundancies. Probably it also can make sense to introduce data classes for the parsing. Anyhow, once I arrive at working on the `Yarn2` code, I would put this ticket on the list of  things to consider.

**Comment by sschuberth** [comment]
> Anyhow, once I arrive at working on the `Yarn2` code, I would put this ticket on the list of things to consider.

I guess it's worth checking whether the issue can still be reproduced as of 0194c8d7c423a0d937b38e5f2d4d0df837517231?

**Comment by sschuberth** [comment]
> The problem can be easily reproduced with an almost empty project and a single dependency:

BTW, I believe that the empty result could be a red herring. If you only have a `package.json` (without a Yarn 2 lock file), then there's no way for ORT to tell for sure it's a Yarn 2 project, so it will fall back to NPM. However, if you disabled NPM (only enabled Yarn 2), this will lead to an empty result.

I'll work on a fix for that particular issue.

**Comment by sschuberth** [comment]
I've experimented with a replacement like

```kotlin
                val dependencies = packageInfo.children.dependencies
                    .map {
                        val a = it.locator
                        val b = a.replace(Regex("@virtual:[0-9a-f]+#(.+)"), "@$1")
                        packageInfoForLocator.getValue(b)
                    }
                    .filter { it.moduleName in dependencyNames }
```

to turn

```
@react-native-community/netinfo@virtual:7ef2ec639f3470d2191275e2f3fd8391c01d875228d71bf82ba9ce9a90858464fb499034540949cf9bef8723b3f2638ad6037a3f99ed4516ef466bf8a24771a3#npm:11.2.1
```

into

```
@react-native-community/netinfo@npm:11.2.1
```

but that gives

```
Exception in thread "main" java.lang.IllegalArgumentException: The following references do not actually refer to packages: [Identifier(type=NPM, namespace=@react-native-community, name=netinfo, version=11.2.1)].
	at org.ossreviewtoolkit.model.utils.DependencyGraphBuilder.checkReferences(DependencyGraphBuilder.kt:204)
```

**Comment by fviernau** [comment]
I propose https://github.com/oss-review-toolkit/ort/pull/10262 as an initial handling of virtual dependencies which includes the dep in the result and no more returns an analyzer issue. I have not investigate how it works with transitive dependencies of virtual dependencies in general.

**Comment by sschuberth** [comment]
> I propose [#10262](https://github.com/oss-review-toolkit/ort/pull/10262) as an initial handling of virtual dependencies which includes the dep in the result and no more returns an analyzer issue.

I've tried your branch with the single-dependency `package.json` file mention in the first post of this issue, but I'm still getting

```
Exception in thread "main" java.lang.IllegalArgumentException: The following references do not actually refer to packages: [Identifier(type=NPM, namespace=@react-native-community, name=netinfo, version=11.2.1)].
	at org.ossreviewtoolkit.model.utils.DependencyGraphBuilder.checkReferences(DependencyGraphBuilder.kt:204)
```

as mentioned [above](https://github.com/oss-review-toolkit/ort/issues/8058#issuecomment-2849270845). Is that expected?

**Comment by fviernau** [comment]
> as mentioned [above](https://github.com/oss-review-toolkit/ort/issues/8058#issuecomment-2849270845). Is that expected?

no it is not. I've tried it myself too, and it seems to work. Can you share your exact reproduce steps?

**Comment by sschuberth** [comment]
> Can you share your exact reproduce steps?

1. Create an empty directory
2. Within that direct, create only a `package.json` file with these contents:

```
{
  "name": "ort-test",
  "packageManager": "yarn@4.0.2",
  "dependencies": {
    "@react-native-community/netinfo": "11.2.1"
  }
}
```

3. In the ORT working space, create a temporary branch based on your `feat-yarn2-virtual-deps` branch, and merge my `node-enabled-fallback` to it (in order to be able to force processing the single `package.json` file with Yarn 2).
4. Run ORT like

```
-Port.forceOverwrite=true -Port.analyzer.enabledPackageManagers=Yarn2 analyze -i /home/sebastian/Downloads/tmp/ -o out/yarn2/analyze
```

where `/home/sebastian/Downloads/tmp/` needs to be adjusted to point to your temporary directory.

5. Observe the log output *correctly* assuming the project to be Yarn 2:

```
Any of [NPM, PNPM, YARN, YARN2] could be the package manager for '/home/sebastian/Downloads/tmp/package.json'. Assuming it is a(n) YARN2 project.
```

6. Observe this exception output directly following

```
Exception in thread "main" java.lang.IllegalArgumentException: The following references do not actually refer to packages: [Identifier(type=NPM, namespace=@react-native-community, name=netinfo, version=11.2.1)].
	at org.ossreviewtoolkit.model.utils.DependencyGraphBuilder.checkReferences(DependencyGraphBuilder.kt:204)
```

**Comment by sschuberth** [comment]
Depending on your Linux setup, you might also need to run `corepack enable` first, to allow bootstrapping of the configured `"packageManager": "yarn@4.0.2"`.

**Comment by fviernau** [comment]
> Depending on your Linux setup, you might also need to run corepack enable first, to allow bootstrapping of the configured "packageManager": "yarn@4.0.2".

I've tested my fix using `project-with-lockfile` which does not bootstrap, but uses `.yarn/releases/yarn-3.2.1.cjs`.

**Comment by fviernau** [comment]
> Maybe. But I believe you'd agree that we should test exactly with the version used by the OP before we can close this issue.

I made another local test. I've replaced `.yarn/releases/yarn-3.2.1.cjs` with `.yarn/releases/yarn-4.0.2.cjs` and updated `plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn2/project-with-lockfile/.yarnrc.yml` accordingly.
The functional test passes. So, I presume that the issue is not related to the handling implemented, but to some differences when bootstrapping is used.

How do you propose to proceed @sschuberth ?

**Comment by sschuberth** [comment]
> How do you propose to proceed [@sschuberth](https://github.com/sschuberth) ?

I have no objections to merging your PR #10262, but IMO the second commit should simply remove the `Resolves: #8058.` part, as we cannot fully prove that yet, it seems.

**Comment by sschuberth** [comment]
> > The problem can be easily reproduced with an almost empty project and a single dependency:
> 
> BTW, I believe that the empty result could be a red herring. If you only have a `package.json` (without a Yarn 2 lock file), then there's no way for ORT to tell for sure it's a Yarn 2 project, so it will fall back to NPM. However, if you disabled NPM (only enabled Yarn 2), this will lead to an empty result.

Actually, there's something else going on: When running `yarn install` on the provided minimal example I get


```
ProcessCapture - Running 'yarn install' in '/home/sebastian/Downloads/tmp'...
ProcessCapture - ➤ YN0000: · Yarn 4.0.2
ProcessCapture - ➤ YN0000: ┌ Resolution step
ProcessCapture - ➤ YN0000: └ Completed
ProcessCapture - ➤ YN0000: ┌ Post-resolution validation
ProcessCapture - ➤ YN0002: │ ort-test@workspace:. doesn't provide react-native (p20ccd), requested by @react-native-community/netinfo.
ProcessCapture - ➤ YN0086: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code.
ProcessCapture - ➤ YN0000: └ Completed
ProcessCapture - ➤ YN0000: ┌ Fetch step
ProcessCapture - ➤ YN0000: └ Completed
ProcessCapture - ➤ YN0000: ┌ Link step
ProcessCapture - ➤ YN0000: └ Completed
ProcessCapture - ➤ YN0000: · Done with warnings in 0s 32ms
```

So Yarn itself already fails in some way (though the exit code is 0), and no `node_modules` directory is being created, which in the end leads to the "The following references do not actually refer to packages" error.

Unfortunately, also `yarn explain peer-requirements p20ccd` doe snot work:

```
The yarn explain peer-requirements command doesn't support this warning type yet.
```

@thirteenthstep, can you maybe conform yourself whether the issue is fixed with the current `main` version of ORT?

**Comment by oheger-bosch** [comment]
@mnonnenmacher , may I ask for the current status of this ticket? Are there already ideas how to fix it or any more concrete time plan? Thanks.

**Comment by mnonnenmacher** [comment]
> [@mnonnenmacher](https://github.com/mnonnenmacher) , may I ask for the current status of this ticket? Are there already ideas how to fix it or any more concrete time plan? Thanks.

For #11208 it is currently only verified that the solution works for direct dependencies, I still need to extend the PR to do the same for transitive dependencies (see the comments in the PR). I would like to do this soon but I cannot tell right now when I will find the time.

**Comment by oheger-bosch** [comment]
Finally, we were able to run a test: While the fix improves the situation, we still see a major discrepancy between the dependencies detected by Yarn classic and Yarn2+ (579 vs 187 in our project). It seems that mainly react and react-native dependencies are affected, for which Yarn2+ does not report any transitive dependencies.

Not sure, should this ticket be reopened or a new one be created?

**Comment by sschuberth** [comment]
> Not sure, should this ticket be reopened or a new one be created?

Please create a new one as the problem does not sound like it's related to virtual packages.

**Comment by oheger-bosch** [comment]
Waiting for feedback from a customer project; then I will probably create a new ticket.

Out of curiosity: Why do you think this is a different issue and not related to virtual packages?

**Comment by sschuberth** [comment]
> Why do you think this is a different issue and not related to virtual packages?

Because you wrote (emphasis mine):

> for which Yarn2+ does not report _any_ transitive dependencies.

And I don't assume that _all_ transitive dependencies (or all direct dependencies one level up) are virtual packages.
